include/zephyr/toolchain/common.h

```diff
@@ -39,6 +39,24 @@
 #endif
 #endif
 
+/*
+ * Thread local variables are declared with different keywords depending on
+ * which C/C++ standard that is used. C++11 and C23 uses "thread_local" whilst
+ * C11 uses "_Thread_local". Previously the GNU "__thread" keyword was used
+ * which is the same in both gcc and g++.
+ */
+#ifndef Z_THREAD_LOCAL
+#if defined(__cplusplus) && (__cplusplus) >= 201103L /* C++11 */
+#define Z_THREAD_LOCAL thread_local
+#elif defined(__STDC_VERSION__) && (__STDC_VERSION__) >= 202311L /* C23 */
+#define Z_THREAD_LOCAL thread_local
+#elif defined(__STDC_VERSION__) && (__STDC_VERSION__) >= 201112L /* C11 */
+#define Z_THREAD_LOCAL _Thread_local
+#else /* Default back to old behavior which used the GNU keyword. */
+#define Z_THREAD_LOCAL __thread
+#endif
+#endif /* Z_THREAD_LOCAL */
```

Comment:
We typically reserve the `Z_*` prefix for internal macros.

Reply:
Ok, makes sense. Didn't find a clear naming convention so I just picked one :). Should I name it like `ZRESTRICT` (it's in the same file) or simply just `THREAD_LOCAL`? Or do you have another suggestion?